common/formatting/layout_optimizer.h

```diff
@@ -0,0 +1,445 @@
+  //    skip or not before_spaces (when merging indent line with normal
+  //    line/layout)
+  // 3. Wrapping layout forces GetSpacesBefore() return 0
+  kLayoutIndent,
+};
+
+std::ostream& operator<<(std::ostream& stream, const LayoutType& type);
+
+// Intermediate partition tree layout
+class Layout {
+ public:
+  Layout(LayoutType type, int spacing)
+      : type_(type), indentation_(0), spaces_before_(spacing) {}
+
+  Layout(const UnwrappedLine& uwline)
+      : type_(LayoutType::kLayoutLine), indentation_(0) {
+    tokens_ = uwline.TokensRange();
+    if (tokens_.size() > 0) {
+      assert(tokens_.size() > 0);
+      spaces_before_ = tokens_.front().before.spaces_required;
+    } else {
+      spaces_before_ = 0;
+    }
+  }
+
+  Layout(int indent)
+      : type_(LayoutType::kLayoutIndent),
+        indentation_(indent),
+        spaces_before_(0) {}
+
+  ~Layout() = default;
+  Layout(const Layout&) = default;
+
+  // Deleting standard interfaces
+  Layout() = delete;
+  Layout(Layout&&) = delete;
+  Layout& operator=(const Layout&) = delete;
+  Layout& operator=(Layout&&) = delete;
+
+  LayoutType GetType() const { return type_; }
+
+  int GetIndentationSpaces() const { return indentation_; }
+
+  int SpacesBeforeLayout() const { return spaces_before_; }
+
+  std::string Text() const {
+    return absl::StrJoin(tokens_, " ",
+                         [=](std::string* out, const PreFormatToken& token) {
+                           absl::StrAppend(out, token.Text());
+                         });
+  }
+
+  UnwrappedLine AsUnwrappedLine() const {
+    assert(type_ == LayoutType::kLayoutLine);
+    UnwrappedLine uwline(0, tokens_.begin());
+    uwline.SpanUpToToken(tokens_.end());
+    uwline.SetPartitionPolicy(PartitionPolicyEnum::kAlwaysExpand);
+    return uwline;
+  }
+
+  int Length() const {
+    assert(type_ == LayoutType::kLayoutLine);
+    if (tokens_.size() == 0) {
+      return 0;
+    }
+    int len = 0;
+    for (const auto& token : tokens_) {
+      len += token.before.spaces_required;
+      len += token.Length();
+    }
+    len -= tokens_.front().before.spaces_required;
+    return len;
+  }
+
+  static int UnwrappedLineLength(const UnwrappedLine& uwline) {
+    int len = 0;
+    for (const auto& token : uwline.TokensRange()) {
+      len += token.before.spaces_required;
+      len += token.Length();
+    }
+    len -= uwline.TokensRange().front().before.spaces_required;
+    return len;
+  }
+
+  SpacingOptions SpacingOptionsLayout() const {
+    assert(type_ == LayoutType::kLayoutLine);
+    assert(tokens_.size() > 0);
+    return tokens_.front().before.break_decision;
+  }
+
+  bool MustWrapLayout() const {
+    return SpacingOptionsLayout() == SpacingOptions::MustWrap;
+  }
+
+  bool MustAppendLayout() const {
+    return SpacingOptionsLayout() == SpacingOptions::MustAppend;
+  }
+
+ private:
+  const LayoutType type_;
+
+  const int indentation_;
+
+  /* const */ FormatTokenRange tokens_;
```

Comment:
What is the commented out `const` indicating ? If it should be const but for whatever reason can not, this should be documented in a comment.

Reply:
`const` uncommented, required minor change in a constructor.